### PR TITLE
Channel

### PR DESCRIPTION
**Pull request by spapinistarkware**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5)

### CODE REVIEW

**Comment by spapinistarkware** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/starkware-libs/stwo-cairo/5?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#5** <a href="https://app.graphite.dev/github/pr/starkware-libs/stwo-cairo/5?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#4** <a href="https://app.graphite.dev/github/pr/starkware-libs/stwo-cairo/4?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3** <a href="https://app.graphite.dev/github/pr/starkware-libs/stwo-cairo/3?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spapinistarkware and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by shaharsamocha7 — CHANGES_REQUESTED** [review]
> Reviewed 2 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5)* status: 4 of 6 files reviewed, 5 unresolved discussions (waiting on @spapinistarkware)

___
*[`stwo_cairo_verifier/src/channel.cairo` line 40 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04baTT31eb36NImNb0:-O04baTT31eb36NImNb1:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L40)):*
> ```Cairo
> pub impl ChannelImpl of ChannelTrait {
>     fn new(digest: felt252) -> Channel {
>         Channel { digest: 0, channel_time: Default::default(), }
> ```





_Suggestion:_
```Cairo
        Channel { digest: digest, channel_time: Default::default(), }
```
___
*[`stwo_cairo_verifier/src/channel.cairo` line 54 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04bwRe2du3qYgoRn8J:-O04bwRe2du3qYgoRn8K:b-v97s2z) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L54)):*
> ```Cairo
> 
>     #[inline]
>     fn draw_base_felts(ref self: Channel) -> [BaseField; FELTS_PER_HASH] {
> ```

add comment/TODO to check that this is sound.



_Code quote:_
```Cairo
    #[inline]
    fn draw_base_felts(ref self: Channel) -> [BaseField; FELTS_PER_HASH] {
```
___
*[`stwo_cairo_verifier/src/channel.cairo` line 88 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04cGpj0Hv7IoIEMcjR:-O04cGpj0Hv7IoIEMcjS:bo6750t) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L88)):*
> ```Cairo
>                                 * M31_IN_HASH_SHIFT
>                                 + x3.into()
>                         );
> ```

Consider to extract to a function.
It is scary because it looks easy to have a bug with this calculation



_Code quote:_
```Cairo
                            ((x0.into() * M31_IN_HASH_SHIFT + x1.into()) * M31_IN_HASH_SHIFT
                                + x2.into())
                                * M31_IN_HASH_SHIFT
                                + x3.into()
                        );
```
___
*[`stwo_cairo_verifier/src/channel.cairo` line 135 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04ciRe5uMV3Hkw5Hdu:-O04ciRe5uMV3Hkw5Hdv:bkjzl3r) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L135)):*
> ```Cairo
>                 break;
>             }
>             let [a, b, c, d, e, f, g, h] = self.draw_base_felts();
> ```

I'm not familiar with those letters :)



_Suggestion:_
```Cairo
let [r0, r1, r2, r3, r4, r5, r6, r7] = self.draw_base_felts();
```
___
*[`stwo_cairo_verifier/src/lib.cairo` line 6 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04bAhh-2b6JtJaVg9u:-O04bAhh-2b6JtJaVg9v:b-tfe412) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/lib.cairo#L6)):*
> ```Cairo
> mod vcs;
> 
> pub use fields::{BaseField, SecureField};
> ```

Why is it here?
maybe you already told me that



_Code quote:_
```Cairo
pub use fields::{BaseField, SecureField};
```

File: stwo_cairo_verifier/src/channel.cairo [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+    fn get_digest(ref self: Channel) -> felt252 {
+        self.digest
+    }
+
+    fn draw_felt252(ref self: Channel) -> felt252 {
+        let res = poseidon_hash_span(array![self.digest, self.channel_time.n_sent.into()].span());
+        self.channel_time.inc_sent();
+        res
+    }
```
Comment (schouhy) [inline stwo_cairo_verifier/src/channel.cairo:51]:
```suggestion
    fn draw_felt252(ref self: Channel) -> felt252 {
        let (s0, _, _) = hades_permutation(self.digest, self.channel_time.n_sent.into(), 2);
        self.channel_time.inc_sent();
        s0
    }
```

File: stwo_cairo_verifier/src/channel.cairo [hunk 2]
```diff
@@ -0,0 +1,252 @@
+use core::array::SpanTrait;
+use core::poseidon::poseidon_hash_span;
```
Comment (schouhy) [inline stwo_cairo_verifier/src/channel.cairo:2]:
```suggestion
use core::poseidon::{poseidon_hash_span, hades_permutation};
```

File: stwo_cairo_verifier/src/channel.cairo [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            extract_m31(ref cur),
+            extract_m31(ref cur),
+            extract_m31(ref cur),
+        ]
+    }
+
+    fn mix_digest(ref self: Channel, digest: felt252) {
+        self.digest = poseidon_hash_span(array![self.digest, digest].span());
+        self.channel_time.inc_challenges();
+    }
```
Comment (schouhy) [inline stwo_cairo_verifier/src/channel.cairo:71]:
```suggestion
    fn mix_digest(ref self: Channel, digest: felt252) {
        let (s0, _, _) = hades_permutation(self.digest, digest, 2);
        self.digest = s0;
        self.channel_time.inc_challenges();
    }
```

File: stwo_cairo_verifier/src/channel.cairo [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+    fn mix_felts(ref self: Channel, mut felts: Span<SecureField>) {
+        let mut res = array![];
```
Comment (schouhy) [inline stwo_cairo_verifier/src/channel.cairo:74]:
```suggestion
        let mut res = array![self.digest];
```

**Review by schouhy — COMMENTED** [review]
Hi! While working on the FRI verifier we had to make these changes to make the channel compatible with the rust implementation.
Let me know what you think!

**Review by spapinistarkware — COMMENTED** [review]
looks good, thanks.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-:-O1Az2Xc3Kk3fy2RfcEf:b6a07hf)* status: 4 of 6 files reviewed, 9 unresolved discussions (waiting on @schouhy and @shaharsamocha7)

___
*[`stwo_cairo_verifier/src/channel.cairo` line 2 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#gh-1667248181:-O1Az3ky5akaaFufqZKN:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L2)):*
<details><summary><i>Previously, schouhy (Sergio Chouhy) wrote…</i></summary><blockquote>

```suggestion
use core::poseidon::{poseidon_hash_span, hades_permutation};
```
</blockquote></details>

Done.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 40 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04baTT31eb36NImNb0:-O1AxWS0BBi0C_ryTPAr:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L40)):*
> ```Cairo
> pub impl ChannelImpl of ChannelTrait {
>     fn new(digest: felt252) -> Channel {
>         Channel { digest: 0, channel_time: Default::default(), }
> ```

Done.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 51 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#gh-1667248011:-O1AxY-BCluBtj2ymDNe:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L51)):*
<details><summary><i>Previously, schouhy (Sergio Chouhy) wrote…</i></summary><blockquote>

```suggestion
    fn draw_felt252(ref self: Channel) -> felt252 {
        let (s0, _, _) = hades_permutation(self.digest, self.channel_time.n_sent.into(), 2);
        self.channel_time.inc_sent();
        s0
    }
```
</blockquote></details>

Done.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 54 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04bwRe2du3qYgoRn8J:-O1AxsF89QGfBcQseoho:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L54)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

add comment/TODO to check that this is sound.
</blockquote></details>

Done.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 71 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#gh-1667248495:-O1Axsl7Bl5G6NTZKCgI:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L71)):*
<details><summary><i>Previously, schouhy (Sergio Chouhy) wrote…</i></summary><blockquote>

```suggestion
    fn mix_digest(ref self: Channel, digest: felt252) {
        let (s0, _, _) = hades_permutation(self.digest, digest, 2);
        self.digest = s0;
        self.channel_time.inc_challenges();
    }
```
</blockquote></details>

Done.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 74 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#gh-1667248638:-O1Axt-16avsY35EWJzd:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L74)):*
<details><summary><i>Previously, schouhy (Sergio Chouhy) wrote…</i></summary><blockquote>

```suggestion
        let mut res = array![self.digest];
```
</blockquote></details>

Done.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 88 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04cGpj0Hv7IoIEMcjR:-O1AyGZK1K_zfIL3PdsB:b-w098yz) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L88)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

Consider to extract to a function.
It is scary because it looks easy to have a bug with this calculation
</blockquote></details>

It just looks scary because of the formatting here IMO, and usage of into.
It's just `((x0 * 2**31 + x1)  * 2**31 + x2) * 2**31 + x3`
___
*[`stwo_cairo_verifier/src/channel.cairo` line 135 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04ciRe5uMV3Hkw5Hdu:-O1Ayi1u8ZZIN26XJfGf:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L135)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

I'm not familiar with those letters :)
</blockquote></details>

Done.
___
*[`stwo_cairo_verifier/src/lib.cairo` line 6 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04bAhh-2b6JtJaVg9u:-O1Ayl4T0ImTvTV9TYoV:b1zvhsh) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/lib.cairo#L6)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

Why is it here?
maybe you already told me that
</blockquote></details>

So it would be easier to import them?
Also, seems ok to export them from the lib.

**Review by shaharsamocha7 — APPROVED** [review]
> Reviewed 1 of 6 files at r1, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5)* status: all files reviewed, 9 unresolved discussions (waiting on @schouhy and @spapinistarkware)

___
*[`stwo_cairo_verifier/src/channel.cairo` line 88 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04cGpj0Hv7IoIEMcjR:-O1B4MMV4ZJRUpbPHOgV:bsayoi0) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L88)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

It just looks scary because of the formatting here IMO, and usage of into.
It's just `((x0 * 2**31 + x1)  * 2**31 + x2) * 2**31 + x3`
</blockquote></details>

but still you use this calculation multiple times.
Wouldn't it be better as an inline function?
___
*[`stwo_cairo_verifier/src/channel.cairo` line 124 at r3](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1B4Wat09wirVplArQv:-O1B4Wat09wirVplArQw:b2kg2ng) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/8e5128ad37b9f3597539857f38051624c5ebb6b4/stwo_cairo_verifier/src/channel.cairo#L124)):*
> ```Cairo
> 
>     fn draw_felt(ref self: Channel) -> SecureField {
>         let [a, b, c, d, _, _, _, _] = self.draw_base_felts();
> ```

r0, .., r3



_Code quote:_
```Cairo
a, b, c, d
```
___
*[`stwo_cairo_verifier/src/channel.cairo` line 222 at r3](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1B4y4GCuYNTcfvv09q:-O1B4y4GCuYNTcfvv09r:b-o1lhz9) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/8e5128ad37b9f3597539857f38051624c5ebb6b4/stwo_cairo_verifier/src/channel.cairo#L222)):*
> ```Cairo
> 
>         // Assert that all the random felts are unique.
> ```

remove extra line
___
*[`stwo_cairo_verifier/src/channel.cairo` line 223 at r3](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1B5-Fn1CJ931W3dIp1:-O1B5-Fn1CJ931W3dIp2:bjoz9w8) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/8e5128ad37b9f3597539857f38051624c5ebb6b4/stwo_cairo_verifier/src/channel.cairo#L223)):*
> ```Cairo
>         // Assert that all the random felts are unique.
> 
>         assert_ne!(random_felts[0], random_felts[5]);
> ```

Consider to create a set(random_felts) and check the lengths are equal



_Code quote:_
```Cairo
assert_ne!(random_felts[0], random_felts[5]);
```
___
*[`stwo_cairo_verifier/src/channel.cairo` line 238 at r3](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1B5EZT7M7-DUEHyBHZ:-O1B5EZT7M7-DUEHyBH_:b-j4b64m) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/8e5128ad37b9f3597539857f38051624c5ebb6b4/stwo_cairo_verifier/src/channel.cairo#L238)):*
> ```Cairo
> 
>         // Reseed channel and check the digest was changed.
>         channel.mix_digest(0);
> ```

move two lines up



_Code quote:_
```Cairo
        channel.mix_digest(0);
```

**Review by spapinistarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5)* status: all files reviewed, 9 unresolved discussions (waiting on @schouhy and @shaharsamocha7)

___
*[`stwo_cairo_verifier/src/channel.cairo` line 88 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O04cGpj0Hv7IoIEMcjR:-O1B8eHL4BDzCh_bmaga:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/f5757e0dbdf12ffdf55eff06818c13a671568a3e/stwo_cairo_verifier/src/channel.cairo#L88)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

but still you use this calculation multiple times.
Wouldn't it be better as an inline function?
</blockquote></details>

Done.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 124 at r3](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1B4Wat09wirVplArQv:-O1B8iTf6y7w097nNrBv:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/8e5128ad37b9f3597539857f38051624c5ebb6b4/stwo_cairo_verifier/src/channel.cairo#L124)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

r0, .., r3
</blockquote></details>

Done.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 222 at r3](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1B4y4GCuYNTcfvv09q:-O1B8lsTAGDtfN8JogLt:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/8e5128ad37b9f3597539857f38051624c5ebb6b4/stwo_cairo_verifier/src/channel.cairo#L222)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

remove extra line
</blockquote></details>

Done.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 223 at r3](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1B5-Fn1CJ931W3dIp1:-O1B8p7U5zEix8vvtNNX:bhzyrre) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/8e5128ad37b9f3597539857f38051624c5ebb6b4/stwo_cairo_verifier/src/channel.cairo#L223)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

Consider to create a set(random_felts) and check the lengths are equal
</blockquote></details>

Kind hard. No set in cairo afaik.
___
*[`stwo_cairo_verifier/src/channel.cairo` line 238 at r3](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1B5EZT7M7-DUEHyBHZ:-O1B9jZmBKPxgPjNlZD9:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/8e5128ad37b9f3597539857f38051624c5ebb6b4/stwo_cairo_verifier/src/channel.cairo#L238)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

move two lines up
</blockquote></details>

Done.

**Review by shaharsamocha7 — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5)* status: 5 of 7 files reviewed, 5 unresolved discussions (waiting on @schouhy and @spapinistarkware)

___
*[`stwo_cairo_verifier/src/utils.cairo` line 76 at r4](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1DMIoz586Z-2aRkQF1:-O1DMIoz586Z-2aRkQF2:b-lrqyo) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/00c67245b2f47e2efe76ab1edebfb2da2b926267/stwo_cairo_verifier/src/utils.cairo#L76)):*
> ```Cairo
> 
> const M31_SHIFT: felt252 = 0x80000000; // 2**31.
> pub fn pack4(cur: felt252, values: [BaseField; 4]) -> felt252 {
> ```

Consider changing to pack8
(or document cur :)



_Code quote:_
```Cairo
pub fn pack4(cur: felt252, values: [BaseField; 4]) -> felt252 {
```

**Review by spapinistarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5)* status: 5 of 7 files reviewed, 5 unresolved discussions (waiting on @schouhy and @shaharsamocha7)

___
*[`stwo_cairo_verifier/src/utils.cairo` line 76 at r4](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1DMIoz586Z-2aRkQF1:-O1G8o5aD9GMcuqSPpdM:b-o0svrk) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/00c67245b2f47e2efe76ab1edebfb2da2b926267/stwo_cairo_verifier/src/utils.cairo#L76)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

Consider changing to pack8
(or document cur :)
</blockquote></details>

Too messy

**Review by shaharsamocha7 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5)* status: 5 of 7 files reviewed, 5 unresolved discussions (waiting on @schouhy and @spapinistarkware)

___
*[`stwo_cairo_verifier/src/utils.cairo` line 76 at r4](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1DMIoz586Z-2aRkQF1:-O1I2SWkAdOxSSO0hMv7:bski6bz) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/00c67245b2f47e2efe76ab1edebfb2da2b926267/stwo_cairo_verifier/src/utils.cairo#L76)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

Too messy
</blockquote></details>

also about cur?
This function name is confusing because of it

**Review by spapinistarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5)* status: 5 of 7 files reviewed, 5 unresolved discussions (waiting on @schouhy and @shaharsamocha7)

___
*[`stwo_cairo_verifier/src/utils.cairo` line 76 at r4](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-O1DMIoz586Z-2aRkQF1:-O1KoL-B3WjCJs6704nM:b-896fix) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/00c67245b2f47e2efe76ab1edebfb2da2b926267/stwo_cairo_verifier/src/utils.cairo#L76)):*
<details><summary><i>Previously, shaharsamocha7 wrote…</i></summary><blockquote>

also about cur?
This function name is confusing because of it
</blockquote></details>

Done.

**Review by shaharsamocha7 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/5#-:-O1L1wh7BL5NW8QO3Mjw:bnfp4nl)* status: 6 of 7 files reviewed, 4 unresolved discussions (waiting on @schouhy)

**Pull request merged**